### PR TITLE
CRUD module with unit tests

### PR DESCRIPTION
**Pull request by keithics**
screen shot 2015-12-26 at 2 39 38 pm

### CODE REVIEW

**Comment by tbaustin** [comment]
Hello, how can we generate a new crud module from cmd line?

**Comment by keithics** [comment]
`yo meanjs:crud-module <module-name>`

**Comment by mbbaig** [comment]
Is this going to be merged anytime soon?

I am using MEANJS 0.4.2 right now and assume that the generator of the same version will soon be released. Since the generators are keeping up with the core MEAN framework.

I need the CRUD generation module. Will this be available in the next release?

**Comment by tbaustin** [comment]
I tried yo meanjs:crud-module and got:

Error meanjs:crud-module customers

You don't seem to have a generator with the name meanjs:crud-module installed.
You can see available generators with npm search yeoman-generator and then install them with npm install [name].
To see the 2 registered generators run yo with the `--help` option.

**Comment by tbaustin** [comment]
@keithics Look at the comment above

**Comment by keithics** [comment]
@tbaustin it hasn't been merged yet.

@mbbaig  @tbaustin you can try my fork for 0.4.2 for the meantime. I have been using it for some time now.
https://github.com/webninjamobile/generator-meanjsplus

**Comment by tbaustin** [comment]
@keithics  How exactly do we use your branch sorry kind of new to a lot of this stuff with node and github.

**Comment by keithics** [comment]
@tbaustin remove the old generator then clone the forked repo. then do `npm link`. You may need to run npm install in order to run it.

http://yeoman.io/authoring/index.html .. see Running the generator

**Comment by tbaustin** [comment]
@keithics  can i just download the zip and run it from that folder ? So I dont have to start all over on my project(I can wait till they get theirs working on it and just make a new one). Then would I just run npm install and then npm link?

**Comment by keithics** [comment]
@tbaustin yes, download.. unzip. cd to that folder. npm install and npm link. if you have the old generator, you need to remove it.

**Comment by tbaustin** [comment]
@keithics How do i remove the old generator ?

File: app/index.js [hunk 1]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (ilanbiala) [inline app/index.js]: Don't remove

File: app/log.js [hunk 2]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (ilanbiala) [inline app/log.js]: Don't remove

File: crud-module/index.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      this.prompt(prompts, function (props) {
+        this.menuId = props.menuId;
+
+        done();
+      }.bind(this));
+    }
+  },
+  createCrudModule: function () {
+    // Create module folder
+    mkdirp('modules/' + this.slugifiedPluralName);
```
Comment (ilanbiala) [inline crud-module/index.js]: You shouldn't need this one, right?

File: crud-module/index.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    mkdirp('modules/' + this.slugifiedPluralName + '/tests');
+
+    // Create module supplemental folders
+    var slugifiedPluralName = this.slugifiedPluralName;
+    this.folders.forEach(function (folder) {
+      mkdirp('modules/' + slugifiedPluralName + '/client/' + folder);
+    })
+
+    // Render express files
+    this.template('server/controllers/_.server.controller.js', 'modules/' + this.slugifiedPluralName + '/server/controllers/' + this.slugifiedPluralName + '.server.controller.js');
```
Comment (ilanbiala) [inline crud-module/index.js]: Can you store these strings in an array and just do a forEach like above?

File: crud-module/index.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    // Render tests files
+    this.template('tests/server/_.server.model.tests.js', 'modules/' + this.slugifiedPluralName + '/tests/server/' + this.slugifiedSingularName + '.server.model.tests.js');
+    this.template('tests/server/_.server.routes.tests.js', 'modules/' + this.slugifiedPluralName + '/tests/server/' + this.slugifiedSingularName + '.server.routes.tests.js');
+    this.template('tests/client/_.client.controller.tests.js', 'modules/' + this.slugifiedPluralName + '/tests/client/' + this.slugifiedPluralName + '.client.controller.tests.js');
+    this.template('tests/e2e/_.e2e.tests.js', 'modules/' + this.slugifiedPluralName + '/tests/e2e/' + this.slugifiedPluralName + '.e2e.tests.js');
+  }
+});
+
+module.exports = CrudGenerator;
```
Comment (ilanbiala) [inline crud-module/index.js]: newline

File: crud-module/templates/client/_.client.module.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+'use strict';
+
+// Use applicaion configuration module to register a new module
+ApplicationConfiguration.registerModule('<%= slugifiedPluralName %>');
```
Comment (ilanbiala) [inline crud-module/templates/client/_.client.module.js]: newline

File: crud-module/templates/client/config/_.client.config.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      state: '<%= slugifiedPluralName %>.list'
+    });
+
+    // Add the dropdown create item
+    Menus.addSubMenuItem('<%= menuId %>', '<%= slugifiedPluralName %>', {
+      title: 'Create <%= humanizedSingularName %>',
+      state: '<%= slugifiedPluralName %>.create'
+    });
+  }
+]);
```
Comment (ilanbiala) [inline crud-module/templates/client/config/_.client.config.js]: newline

File: crud-module/templates/client/config/_.client.routes.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    state('<%= slugifiedPluralName %>.view', {
+      url: '/:<%= camelizedSingularName %>Id',
+      templateUrl: 'modules/<%= slugifiedPluralName %>/client/views/view-<%= slugifiedSingularName %>.client.view.html'
+    }).
+    state('<%= slugifiedPluralName %>.edit', {
+      url: '/:<%= camelizedSingularName %>Id/edit',
+      templateUrl: 'modules/<%= slugifiedPluralName %>/client/views/edit-<%= slugifiedSingularName %>.client.view.html'
+    });
+  }
+]);
```
Comment (ilanbiala) [inline crud-module/templates/client/config/_.client.routes.js]: newline

File: crud-module/templates/client/config/_.client.routes.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+//Setting up route
+angular.module('<%= slugifiedPluralName %>').config(['$stateProvider',
+  function($stateProvider) {
+    // <%= humanizedPluralName %> state routing
+    $stateProvider.
```
Comment (ilanbiala) [inline crud-module/templates/client/config/_.client.routes.js]: need controllers for each state

File: crud-module/templates/client/controllers/_.client.controller.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
+'use strict';
```
Comment (ilanbiala) [inline crud-module/templates/client/controllers/_.client.controller.js]: This file should be changed once we have an example module with the new style guide in the main repo

File: crud-module/templates/client/controllers/_.client.controller.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    };
+
+    // Find existing <%= humanizedSingularName %>
+    $scope.findOne = function() {
+      $scope.<%= camelizedSingularName %> = <%= classifiedPluralName %>.get({
+        <%= camelizedSingularName %>Id: $stateParams.<%= camelizedSingularName %>Id
+      });
+    };
+  }
+]);
```
Comment (ilanbiala) [inline crud-module/templates/client/controllers/_.client.controller.js]: newline

File: crud-module/templates/client/services/_.client.service.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+angular.module('<%= slugifiedPluralName %>').factory('<%= classifiedPluralName %>', ['$resource',
+  function($resource) {
+    return $resource('api/<%= slugifiedPluralName %>/:<%= camelizedSingularName %>Id', { <%= camelizedSingularName %>Id: '@_id'
+    }, {
+      update: {
+        method: 'PUT'
+      }
+    });
+  }
+]);
```
Comment (ilanbiala) [inline crud-module/templates/client/services/_.client.service.js]: newline

File: crud-module/templates/client/views/_.create.client.view.html [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+                <div class="form-group">
+                    <input type="submit" class="btn btn-default">
+                </div>
+				<div data-ng-show="error" class="text-danger">
+					<strong data-ng-bind="error"></strong>
+				</div>
+            </fieldset>
+        </form>
+    </div>
+</section>
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.create.client.view.html]: newline

File: crud-module/templates/client/views/_.create.client.view.html [hunk 14]
```diff
@@ -0,0 +1,23 @@
+<section data-ng-controller="<%= classifiedPluralName %>Controller">
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.create.client.view.html]: Needs style guide update

File: crud-module/templates/client/views/_.edit.client.view.html [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+                <div class="form-group">
+                    <input type="submit" value="Update" class="btn btn-default">
+                </div>
+				<div data-ng-show="error" class="text-danger">
+					<strong data-ng-bind="error"></strong>
+				</div>
+            </fieldset>
+        </form>
+    </div>
+</section>
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.edit.client.view.html]: newline

File: crud-module/templates/client/views/_.edit.client.view.html [hunk 16]
```diff
@@ -0,0 +1,23 @@
+<section data-ng-controller="<%= classifiedPluralName %>Controller" data-ng-init="findOne()">
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.edit.client.view.html]: style guide update

File: crud-module/templates/client/views/_.list.client.view.html [hunk 17]
```diff
@@ -0,0 +1,19 @@
+<section data-ng-controller="<%= classifiedPluralName %>Controller" data-ng-init="find()">
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.list.client.view.html]: style guide update

File: crud-module/templates/client/views/_.list.client.view.html [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+				by
+				<span data-ng-bind="<%= camelizedSingularName %>.user.displayName"></span>
+			</small>
+			<h4 class="list-group-item-heading" data-ng-bind="<%= camelizedSingularName %>.name"></h4>
+        </a>
+    </div>
+    <div class="alert alert-warning text-center" data-ng-hide="!<%= camelizedPluralName %>.$resolved || <%= camelizedPluralName %>.length">
+    	No <%= humanizedPluralName %> yet, why don't you <a href="/#!/<%= slugifiedPluralName %>/create">create one</a>?
+    </div>
+</section>
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.list.client.view.html]: newline

File: crud-module/templates/client/views/_.view.client.view.html [hunk 19]
```diff
@@ -0,0 +1,21 @@
+<section data-ng-controller="<%= classifiedPluralName %>Controller" data-ng-init="findOne()">
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.view.client.view.html]: style guide update

File: crud-module/templates/client/views/_.view.client.view.html [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+	</div>
+	<small>
+		<em class="text-muted">
+			Posted on
+			<span data-ng-bind="<%= camelizedSingularName %>.created | date:'mediumDate'"></span>
+			by
+			<span data-ng-bind="<%= camelizedSingularName %>.user.displayName"></span>
+		</em>
+	</small>
+</section>
```
Comment (ilanbiala) [inline crud-module/templates/client/views/_.view.client.view.html]: newline

File: crud-module/templates/server/controllers/_.server.controller.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * <%= humanizedSingularName %> middleware
+ */
+exports.<%= camelizedSingularName %>ByID = function(req, res, next, id) { <%= classifiedSingularName %>.findById(id).populate('user', 'displayName').exec(function(err, <%= camelizedSingularName %>) {
+  if (err) return next(err);
+  if (! <%= camelizedSingularName %>) return next(new Error('Failed to load <%= humanizedSingularName %> ' + id));
+  req.<%= camelizedSingularName %> = <%= camelizedSingularName %> ;
+  next();
+});
+};
```
Comment (ilanbiala) [inline crud-module/templates/server/controllers/_.server.controller.js]: newline

File: crud-module/templates/server/controllers/_.server.controller.js [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    res.jsonp(<%= camelizedPluralName %>);
+  }
+});
+};
+
+/**
+ * <%= humanizedSingularName %> middleware
+ */
+exports.<%= camelizedSingularName %>ByID = function(req, res, next, id) { <%= classifiedSingularName %>.findById(id).populate('user', 'displayName').exec(function(err, <%= camelizedSingularName %>) {
+  if (err) return next(err);
```
Comment (ilanbiala) [inline crud-module/templates/server/controllers/_.server.controller.js]: braces need on all ifs, even if they can be single liners

File: crud-module/templates/server/models/_.server.model.js [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    type: Date,
+    default: Date.now
+  },
+  user: {
+    type: Schema.ObjectId,
+    ref: 'User'
+  }
+});
+
+mongoose.model('<%= classifiedSingularName %>', <%= classifiedSingularName %>Schema);
```
Comment (ilanbiala) [inline crud-module/templates/server/models/_.server.model.js]: newline

File: crud-module/templates/server/routes/_.server.routes.js [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+    .post(<%= camelizedPluralName %>.create);
+
+  app.route('/api/<%= slugifiedPluralName %>/:<%= camelizedSingularName %>Id').all(<%= camelizedPluralName %>Policy.isAllowed)
+    .get(<%= camelizedPluralName %>.read)
+    .put(<%= camelizedPluralName %>.update)
+    .delete(<%= camelizedPluralName %>.delete);
+
+  // Finish by binding the <%= humanizedSingularName %> middleware
+  app.param('<%= camelizedSingularName %>Id', <%= camelizedPluralName %>.<%= camelizedSingularName %>ByID);
+};
```
Comment (ilanbiala) [inline crud-module/templates/server/routes/_.server.routes.js]: newline

File: crud-module/templates/tests/client/_.client.controller.tests.js [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+
+      // Run controller functionality
+      scope.remove(sample<%= classifiedSingularName %>);
+      $httpBackend.flush();
+
+      // Test array after successful delete
+      expect(scope.<%= camelizedPluralName %>.length).toBe(0);
+    }));
+  });
+}());
```
Comment (ilanbiala) [inline crud-module/templates/tests/client/_.client.controller.tests.js]: newline

File: package.json [hunk 26]
```diff
@@ -32,6 +32,13 @@
   "dependencies": {
     "bluebird": "~3.0.6",
     "chalk": "~1.1.1",
+    "ejs": "^2.3.4",
```
Comment (ilanbiala) [inline package.json]: all should use ~ and should be latest versions

File: test/app.test.js [hunk 27]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (ilanbiala) [inline test/app.test.js]: don't remove

File: test/crud-module.test.js [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+      assert.file('modules/foos/tests/e2e/foos.e2e.tests.js');
+    });
+
+
+
+
+  });
+
+
+});
```
Comment (ilanbiala) [inline test/crud-module.test.js]: newline

File: test/crud-module.test.js [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+      assert.file('modules/foos/client/config/foos.client.config.js');
+    });
+
+    it('should generate an express server test files', function () {
+      assert.file('modules/foos/tests/server/foo.server.model.tests.js');
+      assert.file('modules/foos/tests/server/foo.server.routes.tests.js');
+      assert.file('modules/foos/tests/client/foos.client.controller.tests.js');
+      assert.file('modules/foos/tests/e2e/foos.e2e.tests.js');
+    });
+
```
Comment (ilanbiala) [inline test/crud-module.test.js]: too many

File: test/crud-module.test.js [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+
+
+
+  });
+
```
Comment (ilanbiala) [inline test/crud-module.test.js]: too many

**Comment by ilanbiala** [comment]
@keithics did an initial review. It looks pretty good, though the style of the code needs to be updated once we finalize a couple modules in the main repo that you can refer to and copy from.

@codydaig did I miss anything in the initial review?

**Comment by keithics** [comment]
thanks @ilanbiala let me know once the modules are finalized.

**Comment by tbaustin** [comment]
@keithics When I type grunt in the cmd it does not work. I just made a completely new project and cloned your fork using git clone ~.

**Comment by jpimentel** [comment]
Still waiting for this

**Comment by codydaig** [comment]
@keithics Most of the modules have been merged in, and the remaining few are almost done. You can look at those for styles and what not. Let us know when you've addressed the above comments from @ilanbiala and resolve the conflicts and we can go into a more detailed review.

**Comment by keithics** [comment]
@codydaig will take a look at it now. thanks.

File: crud-module/index.js (hunk 3 above)
Comment (keithics) [inline crud-module/index.js]: yes, Create module folder is from the old one but the menu seems needed if you have another menu other than topbar.

File: crud-module/templates/client/config/_.client.routes.js (hunk 9 above)
Comment (keithics) [inline crud-module/templates/client/config/_.client.routes.js]: sorry, i don't get this note. Can you explain? thanks.

**Comment by trendzetter** [comment]
will this build new style modules when I clone it to test it now? this could be very useful for me right now.

**Comment by keithics** [comment]
@trendzetter please feel free to test it..

**Comment by trendzetter** [comment]
I was about to and then I realized that this would still produce "old style" modules. I am thinking about updating it to make it produce new modules, seems quite a bit of change though. But setting op modules manually is also laborious

**Comment by UndeadBaneGitHub** [comment]
@trendzetter CRUD generator for new style (both folders AND codestyle itself) is in https://github.com/meanjs/generator-meanjs/pull/201

**Comment by codydaig** [comment]
@keithics Thanks for the hard work! We decided to merge @UndeadBaneGitHub's PR because it matched the latest style guide. Again, thanks for the hard work, and I hope to see more PRs from you in the future!

**Comment by keithics** [comment]
awesome! No worries guys.

**Pull request closed without merging**